### PR TITLE
HIVE-25907: IOW Directory queries fails to write data to final path when query result cache is enabled

### PR DESCRIPTION
**Pull request by shameersss1**
### Why are the changes needed?
INSERT OVERWRITE DIRECTORY queries fails to write the data to the specified directory location when query result cache is enabled. This issue is related to HIVE-21386 where query result cache is ignored for insert into, insert overwrite table type queries where as insert overwrite directory case is missed. So this commit fixes that gap by not using query result cache when query is of type insert overwrite directory.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
New tests were added to validate the fix

### CODE REVIEW

**Comment by shameersss1** [comment]
@kgyrtkirk @vineetgarg02 Could you please review the changes?

**Comment by shameersss1** [comment]
@kgyrtkirk - Could you please review the changes?
Thanks

**Comment by shameersss1** [comment]
@kgyrtkirk  - Can you please review the changes?

**Comment by shameersss1** [comment]
@kgyrtkirk - Can you please review the changes?

**Comment by shameersss1** [comment]
@kgyrtkirk - Can you please review the changes?

**Comment by shameersss1** [comment]
@kgyrtkirk @rbalamohan Could you please review the changes?

**Comment by shameersss1** [comment]
@kgyrtkirk - Can you please review the changes?

**Comment by shameersss1** [comment]
@kgyrtkirk Could you please review the changes?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/QBParseInfo.java [hunk 1]
```diff
@@ -687,6 +694,10 @@ public boolean hasInsertTables() {
     return this.insertIntoTables.size() > 0 || this.insertOverwriteTables.size() > 0;
   }
 
+  public boolean hasInsertOverwriteDirectory() {
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/QBParseInfo.java:697]: nit: `isInsertOverwriteDirectory`

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 2]
```diff
@@ -1681,6 +1681,9 @@ boolean doPhase1(ASTNode ast, QB qb, Phase1Ctx ctx_1, PlannerContext plannerCtx)
               && ch.getChild(0) instanceof ASTNode) {
             ch = (ASTNode) ch.getChild(0);
             isTmpFileDest = ch.getToken().getType() == HiveParser.TOK_TMP_FILE;
+            if (ch.getToken().getType() == HiveParser.StringLiteral) {
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:1684]: Could this be something else?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/QBParseInfo.java (hunk 1 above)
Comment (shameersss1) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/QBParseInfo.java:697]: ack

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 2 above)
Comment (shameersss1) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:1684]: I don't think so. As per my understanding "DIRECTORY" is always followed by string literal and this happens only for INSERT query.

**Comment by shameersss1** [comment]
@pvary - Looks like pre-commit is broken. I will re-trigger the tests again. Are we good with the changes?

**Comment by shameersss1** [comment]
@pvary - Tests passed on re-run. Are we good to take it forward?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 2 above)
Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:1686]: Then, I think this should be an assert.

Comment (shameersss1) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:1686]: why assert? The code block inside the condition `if (ast.getChildCount() > 0 && ast.getChild(0) instanceof ASTNode) {
          ASTNode ch = (ASTNode) ast.getChild(0);`  is also accessed by other queries for example queries which have `HiveParser.TOK_TMP_FILE  `

Comment (shameersss1) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:1686]: @pvary 
For Example : `ANALYZE TABLE src COMPUTE STATISTICS FOR COLUMNS key,value` such queries will hit the above condition and having assert will break it right? Is my understanding correct or am i missing anything?

Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:1686]: What will be the value for `ch.getToken().getType()` in this case?

Comment (pvary) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:1686]: Ahh, so it could be something else then 😄 
That explains this line, thanks

**Review by pvary — APPROVED** [review]

**Comment by shameersss1** [comment]
@pvary - Thanks for the review. 
Are we good to merge this PR?

**Pull request merged**